### ISSUE TITLE
Type inference for async generators

### ISSUE THREAD

**Issue by dimaqq** [microsoft/TypeScript#44264]
# Suggestion

It would be amazing if typescript could infer types from async generators.
The following is valid javascript, thus, I think, it should be valid typescript and type of `x` should be inferred.
Today it fails with `Parameter 'x' implicitly has 'any' type.`

```
async function forever(x) {
  for await (const unused of x) {
    void unused;
  }
}

async function* test() {
  for (;;) {
    await something();
    yield;
  }
}

forever(test());
```

## 🔍 Search Terms

async generator, async iterator, type inference, infer types

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Treat type of `x` as "any async iterator would do" thus including any async generator function. 

## 📃 Motivating Example

My own use... forever/background (async) processes/

## 💻 Use Cases

Today, I need to type this explicitly, which is unnecessarily verbose.

In the long run, it would allow **composing** async functions same as type inference helps composing plain functions today.

I get it that type inference for async iterables/iterators/generators opens a huge can of worms, like:
* type narrowing, like #36687 or in this vein
* yield expression return type #36967
* generator yielded type #32523

**Comment by DanielRosenwasser** [comment]
This sounds like use-site based inference, which is discussed more generally over at #15114. It would be awfully strange for us to infer the types of iterable values, async values, or especially just async iterable variables.

**Comment by DanielRosenwasser** [comment]
> that is an old ticket that is closed and locked...

Right, what I mean is that it's very unlikely that we will be pursuing that direction, and many of the same issues would be relevant here. Even if we could solve it for a subset, it would be a pretty inconsistent ruleset of inference.

**Comment by nightlyherb** [comment]
Could something like this possibly work?
Apologies in advance if I lack a basic understanding of something fundamental...
https://www.typescriptlang.org/play?#code/PTAEBUAsEsGdQCYHsCm8B2SAuoDGT0sAnAQwWlxy0hVCwE8AHW2e2LFAWwBpRIlmAMwCuAG1H0A-AChp0QiiKCSuWgFkCZCgHEU6RSSxIiAHgCa0FKIQBhJMMKgUADw7oEGYZwBGi3hasEABk4HBc3D1AHAGtMAHd0AG0AXV4AJRQsYSJ0J1c9SJJ0el4AOXCAQSIAc1gQ9jyI+Bj4pOSUgD5QAG9pUH7QfVcTAEl3F0aC+ADrOwcsDoAKADpVkhrYAC5Qctcq2vqsRLGEF2SASm2RjlIjIgysnIzYMSwTB+z0DoBuPoGht4nCbhKaDLy+IigAC8YJ8iiWq2W61q212WH2dVCx3GzguVxuhmMMwQz1e5ks1kO2NOuJ+0gAvrJ8OgGtU9AAxBy4aGgEToSjQAgAKlAi3OPT+-WZDRIPPoFIQgxQcVAAClhOxFgBWc6-AZ4AgNbxyhVKlXqzUAIgAbJbdZLQERMp8zWqNVhFrKANSgRjrWAoMYe7y8ACMAAZzvb6aASPAxdCuhp0FpcLp9LdjCYAEy8RJqEj0XwmdDg+G8AtFlAmdhEeTVDqpUCV4uluFEDp5xIOU6CeQoBBNxJtiFD2v15LJDqyPkCgiDEjQABuKAAIkhSthDIL0CZwOHeOBQ+kuosHWz0Jz+dsE1Ck5pyGm9AY7jm8y3q-vO83C8Wj42K1-as0m-RJu3GPt9EHPN9yHI9J2ncVen1aUcAvHkLyvXAxT1AZUJ6UAlxIURhBQbZOFlGMYQvZYARMcMlh7FBIIHe19SdR5cgo5ZvHkBBFk9cU7wlfUUMNHBukI4jSPI40qNAGi6NDJYSDY0T+g4l0fB4viBO8ISumQ9TRPwySiJIsjHTQV5QHkxTwhzJZ9Nw4yBk0nIrJeUQsBc0T6XOB1-IZWQQB-U5Y1ATg0HgYh5XQao6CQBTn1IURoAAL1oah7GqSAQrAdkiCQTg+CwLBGC2EBcGyQVYGgBgAFoEDrFd0GWYxqmATgH1gBr5AagArEgiNgXA60YLAAGJkHy5sH1kBhmDmlMED3LoYSM-pePcEwAFUlkEG8sG2cADOWsg9o6S5ztW-bgukRb1AfMYM3AJhqzUcBJnceBkwuop6A6dbmy+kEfpukx5EERQIC6SQIFAbZ9BXIhflmj8mVEON4AtN5wC6aBOEYUQuD0LBfofNaRIGcyZIgNGxJZYhhEoYxFlpyzTup9TqDgZYOZ5DnfMZB1ttug6bw5k6zr+8Xrtly7udE9zckERZedgfnpJQNT+kZfUHSMABlYh6wTTbledDyAANccWAASboNa1iz-Ot4X7vwzdefinkLbFy7FkO0UpaVFGZcp-b5cjwyHXYq3cg14XuFkfVjdN+LzbjtyE9AS1vZgeLLWT+7Hp-KsqZhXGqYAHzod6kEEHZsEL6pZAe97QDGMbScIUZvsiEd4R5B1EmPUBc1AABmXgABZeC1XhrV4AB2XgAA5eAATjDA9QFDcfQ0n0MZ-3+f98X-fl-3tf983-ed4nvfs3H7NJ+zU-s3P7NL8RIeiBSO0EYyRfhl1XCgHuUU+4jAHp4ds61R4NXHnvV+vBT7n0vtfW+99H4RjDIfY+p9Qzn1DJfUM19Qy31DPfUMj9szP1fu-T+58-5lgAZOY4ICFqdwLIwBWoN8jgwBp0P2oBEjQFAPIUA0QUD0EbhAZI2x+HiKnIyMuvDwAbkqBsPcsDYzFBERtMREipEyLkU3cAijJF6P-qAeGiRwCJHAZAsmkNGzJERvo+gtkZxciwDuEUyBvbbgICYB0xI5iODBoPNhKd9QADVtaHD0S0JACQUhxIGB8HIb1mDSFPJhLkN4I4rR0ClQkphwkKkiVgTJ-ReEK0SRZQ4nYHQK2yegXJKBWn6g0VovYOimmkRafkpCDp8LoWohyLkOEHSzn8fOdgKBGD92iXAiESxQ5DJQFSZxTooGAkbNHFaJhtlUmAbHVy+EnReRwFMtqdERibO1rrUS0Am6LBua8ZYyB9BjNcvHTinlvlC2znrMFVkgVfO8i7UiOkdpnKxN3fZrinlTgEvIDM2yzpLJWci3uhyliYsUNi15+sc5AtxfRRiEF+wIGjEycSwLvI8iCVuBZ6BFiFP5PaIAA
